File: backend/src/services/auto-develop-status.ts
/**
 * Shared helper for assembling AutoDevelopStatus.
 *
 * Used by both the REST route (GET /api/projects/:id/auto-develop)
 * and the MCP tool (get_auto_develop_status) to avoid duplication (adj-122.10.6).
 */

import type { AutoDevelopStatus } from "../types/auto-develop.js";
import type { ProposalStore } from "./proposal-store.js";
import type { AutoDevelopStore } from "./auto-develop-store.js";

interface ProjectForStatus {
  id: string;
  name: string;
  autoDevelop: boolean;
  autoDevelopPausedAt?: string | null | undefined;
  visionContext?: string | null | undefined;
  autoDevelopProductOwner?: string | null | undefined;
}

/**
 * Build an AutoDevelopStatus object from project data and stores.
 */
export function buildAutoDevelopStatus(
  project: ProjectForStatus,
  proposalStore: ProposalStore | undefined,
  autoDevelopStore: AutoDevelopStore | undefined,
): AutoDevelopStatus {
  const pendingProposals = proposalStore?.getProposals({ status: "pending", project: project.id }) ?? [];
  const escalated = pendingProposals.filter(
    p => p.confidenceScore !== undefined && p.confidenceScore >= 40 && p.confidenceScore < 60,
  ).length;
  const inReview = pendingProposals.length - escalated;
  const accepted = proposalStore?.getProposals({ status: "accepted", project: project.id }).length ?? 0;
  const dismissed = proposalStore?.getProposals({ status: "dismissed", project: project.id }).length ?? 0;

  const activeCycle = autoDevelopStore?.getActiveCycle(project.id) ?? null;
  const cycleHistory = autoDevelopStore?.getCycleHistory(project.id) ?? [];
  const completedCycles = cycleHistory.filter(c => c.completedAt !== null).length;

  return {
    enabled: project.autoDevelop,
    paused: !!project.autoDevelopPausedAt,
    pausedAt: project.autoDevelopPausedAt ?? null,
    currentPhase: activeCycle ? (activeCycle.phase as AutoDevelopStatus["currentPhase"]) : null,
    activeCycleId: activeCycle?.id ?? null,
    visionContext: project.visionContext ?? null,
    proposals: { inReview, accepted, escalated, dismissed },
    epicsInExecution: accepted,
    cycleStats: {
      totalCycles: cycleHistory.length,
      completedCycles,
      currentCycleNumber: cycleHistory.length > 0 ? cycleHistory.length : 0,
    },
    productOwner: project.autoDevelopProductOwner ?? null,
  };
}

File: ios/AdjutantKit/Sources/AdjutantKit/Models/AutoDevelopStatus.swift
import Foundation

/// Status of the auto-develop loop for a project.
/// Mirrors the backend auto-develop status response.
public struct AutoDevelopStatus: Codable, Equatable {
    public let enabled: Bool
    public let paused: Bool
    public let pausedAt: String?
    public let currentPhase: String?
    public let activeCycleId: String?
    public let visionContext: String?
    public let proposals: ProposalCounts
    public let epicsInExecution: Int
    public let cycleStats: CycleStats

    public init(
        enabled: Bool,
        paused: Bool,
        pausedAt: String? = nil,
        currentPhase: String? = nil,
        activeCycleId: String? = nil,
        visionContext: String? = nil,
        proposals: ProposalCounts,
        epicsInExecution: Int,
        cycleStats: CycleStats
    ) {
        self.enabled = enabled
        self.paused = paused
        self.pausedAt = pausedAt
        self.currentPhase = currentPhase
        self.activeCycleId = activeCycleId
        self.visionContext = visionContext
        self.proposals = proposals
        self.epicsInExecution = epicsInExecution
        self.cycleStats = cycleStats
    }

    /// Counts of proposals in various states.
    public struct ProposalCounts: Codable, Equatable {
        public let inReview: Int
        public let accepted: Int
        public let escalated: Int
        public let dismissed: Int

        public init(inReview: Int, accepted: Int, escalated: Int, dismissed: Int) {
            self.inReview = inReview
            self.accepted = accepted
            self.escalated = escalated
            self.dismissed = dismissed
        }
    }

    /// Cycle completion statistics.
    public struct CycleStats: Codable, Equatable {
        public let totalCycles: Int
        public let completedCycles: Int
        /// Current cycle number (1-based). Falls back to totalCycles if not provided.
        public let currentCycleNumber: Int

        public init(totalCycles: Int, completedCycles: Int, currentCycleNumber: Int? = nil) {
            self.totalCycles = totalCycles
            self.completedCycles = completedCycles
            self.currentCycleNumber = currentCycleNumber ?? totalCycles
        }
    }
}
